### PR TITLE
Enhancement/issue 14/rss feed

### PR DESCRIPTION
**Pull request by bakhterets**
Need to clarify additional requirements

### CODE REVIEW

**Comment by bakhterets** [comment]
image
Now it looks like this

File: app/__init__.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def create_app(test_config=None):
     ):
         app.config["SQLALCHEMY_DATABASE_URI"] = app.config["DATABASE_URI"]
     if "SQLALCHEMY_DATABASE_URI" not in app.config:
-        # TODO(gtema): sooner or later this should be dropped
+        # TODO(gtema): sooner or later this should be droppedy
```
Comment (gtema) [inline app/__init__.py:61]: why this change?

File: app/api/schemas/components.py [hunk 2]
```diff
@@ -30,8 +30,8 @@ class IncidentSchema(Schema):
     id = fields.Integer(dump_only=True)
     text = fields.String(required=True)
     impact = fields.Integer(required=True)
-    start_date = fields.Date(required=True)
-    end_date = fields.Date()
+    start_date = fields.Date(format="%Y-%m-%d %H:%M:%S", required=True)
```
Comment (gtema) [inline app/api/schemas/components.py:33]: lets drop seconds precision, this is useless

File: app/rss/routes.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    )
+    if not component:
+        return "Component not found", 404
+    component_data = ComponentSchema().dump(component)
+    incidents = component_data["incidents"]
+    none_end_date_list = []
+    sorted_list = []
+    limited_incidents = []
+    for incident in incidents:
+        if incident["end_date"] is None:
```
Comment (gtema) [inline app/rss/routes.py:41]: I do not get the idea here. Why do you do this?

File: app/rss/routes.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    sorted_list = sorted_list[:9]
+    limited_incidents.extend(none_end_date_list)
+    limited_incidents.extend(sorted_list)
+    limited_incidents = limited_incidents[:10]
+    #
+    # Create feed generator
+    #
+    fg = FeedGenerator()
+    fg.title(f"{component_name} ({region}) - Incidents")
+    fg.link(href=f"http://localhost/rss/{region}/{component_name}", rel="self")
```
Comment (gtema) [inline app/rss/routes.py:61]: localhost is definitely wrong here

File: app/rss/routes.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                end_date = None
+            content_dict = {
+                "type": "text/html",
+                "value": (
+                    f"Impact: {incident['impact']}<br>"
+                    f"Start date: {start_date_str}<br>"
+                    f"End date: {end_date_str if end_date else ''}"
+                )
+            }
+            content_string = json.dumps(content_dict)
```
Comment (gtema) [inline app/rss/routes.py:88]: It is required to urlencode the content if it is not a plain text. On the other side putting json into the html into xml is also not a great idea. In my opinion a plain "text" is enough, no more details are necessary. But the text should be not simply the "Incident" const, but a text string of the incidentupdate.

Something in the form:
```python
for incident in <last_10_incidents>:
  rss.add_entry(title=incident.text, description=f"{incident.impact} incident affecting components: {incident.components}", pubDate=incident.start_date)
  for update in incident.updates:
    rss.add_entry(title=f"{incident.text} update", description=update.text, pubDate=update.timestamp)
```

**Review by gtema — CHANGES_REQUESTED** [review]

File: app/rss/routes.py (hunk 4 above)
Comment (bakhterets) [reply app/rss/routes.py:61]: yep, need to change to variable with address or name

File: app/rss/routes.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                                        Update: {update['text']}, \
+                                        Update timestamp: {update_timestamp} \
+                                        </div>"
+                    fe.pubDate(update_timestamp)
+                else:
+                    fe.pubDate(start_date)
+                fe.content(f"<div class='item_desc'>{content_string}</div>")
+
+        rss_string = fg.rss_str(pretty=True).decode("utf-8")
+        rss_string = rss_string.replace("<span", "<div")
```
Comment (gtema) [inline app/rss/routes.py:184]: I do not think it is a good idea to do such replacements in the content generated by 3rd party tool outside of control. It should not be necessary also.

File: app/rss/routes.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            component_data = ComponentSchema().dump(component)
+            incidents.extend(component_data["incidents"])
+
+        for incident in reversed(sorted_incidents(incidents)):
+            if incident["end_date"] is None or datetime.strptime(
+                incident["end_date"], "%Y-%m-%d %H:%M"
+            ) <= datetime.today():
+                fe = fg.add_entry()
+                fe.title(incident["text"])
+                start_date = tz.localize(
```
Comment (gtema) [inline app/rss/routes.py:140]: pulling additional dependency is not necessary. Same can be achieved as
```
start_date = incident["start_date"].astimezone()
```

**Review by gtema — CHANGES_REQUESTED** [review]

File: app/api/routes.py [hunk 8]
```diff
@@ -141,7 +142,7 @@ def post(self, data):
             # No active incidents - open new
             current_app.logger.debug("No active incidents - opening new one")
             new_incident = Incident(
-                text="Incident",
+                text=text if text is not None else "Incident",
```
[the remark this answers is not in the document]
Comment (bakhterets) [reply app/api/routes.py:145]: It will be changed to: 
text = data.get("text", "Incident")

File: app/rss/routes.py (hunk 3 above)
Comment (bakhterets) [reply app/rss/routes.py:33]: First thought was because I wanted to add an always active Incident to the output to show if it was changed manually through the request to the database which is contrary to the design of the app. It's not helpful maybe but I thought it would be helpful

**Comment by gtema** [comment]
Screenshot 2023-05-25 at 17 02 09

**Review by gtema — APPROVED** [review]
